### PR TITLE
Coalesce build caches

### PR DESCRIPTION
**Pull request by scinos**
### Background

We have a few caches all over the place:
* TypeScript caches are in `.tsc-cache`
* css-loader caches are in `.cache/css`
* Babel caches are in `./build/.babel-client-cache`
* Moment caches are in `./build/.moment-timezone-data-webpack-plugin-cache`

### Changes

This PR moves all those caches into directories inside `./cache` for easier cleanup

### Testing instructions

* Verify tests passes
* Checkout the branch and run `yarn start` two times. Verify `.cache` has some content, and the second time was faster than the first one.

### CODE REVIEW

File: .circleci/config.yml
Comment (scinos) [inline .circleci/config.yml:180]: This was not used 🤷

File: packages/calypso-build/webpack/sass.js
Comment (nsakaimbo) [inline packages/calypso-build/webpack/sass.js:34]: 👋 This will break caching for wp-desktop. The desktop's CI config in config.yml will need to be updated (and tested) for this change as well. Desktop cache-loader saving and restoration is located [here](https://github.com/Automattic/wp-calypso/blob/7d9ea1f8227eb780c89acce1854bc3350a11fd38/.circleci/config.yml#L232-L240).

Once this logic is updated, we'd want to apply "Re-run from start" to the wp-desktop-source build to ensure that the updated cache paths are  re-built/saved and restored correctly (and that optimized build times are not affected).

File: packages/calypso-build/webpack/sass.js
Comment (nsakaimbo) [inline packages/calypso-build/webpack/sass.js:36]: Also, this cache-loader path is used in `composite-checkout` and will need to be updated [here](https://github.com/Automattic/wp-calypso/blob/7d9ea1f8227eb780c89acce1854bc3350a11fd38/packages/composite-checkout/webpack.config.js#L20).

**Review by nsakaimbo — DISMISSED** [review]
Thanks for tackling this, @scinos! I left a couple notes on items that we should take a look at before landing this.

File: packages/calypso-build/webpack/sass.js
Comment (nsakaimbo) [reply packages/calypso-build/webpack/sass.js:34]: General question: why is `cache-loader` being completely dropped here? Based on recent findings, this optimization is significant.

Comment (nsakaimbo) [reply packages/calypso-build/webpack/sass.js:34]: Follow-up: there are other parts of the CI configuration that depend on cache-loader specifically, and it's unclear how removing/replacing `cache-loader` will work. Specifically, cache-loader uses file modified time to ensure consistency between builds. [This part of the Circle config](https://github.com/Automattic/wp-calypso/blob/master/.circleci/config.yml#L686-L696) will have to be revisited as well.

Comment (scinos) [reply packages/calypso-build/webpack/sass.js:34]: It is not removed, it has been deduplicated.

Before this change, `cache-loader` was loaded twice:

image

The change in this PR removes the second instance. Now `cache-loader` is only added if there is a `cacheDirectory` in the options, so the consumer can specify where they want the cache and this lib doesn't have to make any assumptions as where that cache will be located.

We could add a default here, but that will remove the control from the consumer. This way (without the default), the consumer can either enable the cache and have control over the cache path, or disable the cache (by no providing a cache path).

**Review by sirreal — COMMENTED** [review]
I haven't tested, but I think the tsc cache changes should be fine 👍 

**EDIT:** See my comment https://github.com/Automattic/wp-calypso/pull/44863#discussion_r476284289

File: packages/calypso-build/webpack/sass.js
Comment (scinos) [reply packages/calypso-build/webpack/sass.js:34]: I updated CircleCI config to cache the whole `./cache` (not only the CSS). In theory it should be a bit faster now, I'll run the build a few times to get numbers so we can compare with master.

Comment (scinos) [reply packages/calypso-build/webpack/sass.js:34]: With the new cache, when the cache is full it takes a bit above 8 minutes to build the source:

image

Which seems to be at least ~4 minutes faster than master:

image

Comment (sirreal) [reply packages/calypso-build/webpack/sass.js:34]: > I updated CircleCI config to cache the whole ./cache (not only the CSS).

This has the potential to break with TypeScript. If TypeScript finds valid cache files, it won't do any work. That means that the build output will not be produced under some circumstances if only the cache is persisted. We need to make sure that both the output and the cache are persisted, or neither.

Comment (scinos) [reply packages/calypso-build/webpack/sass.js:34]: Oh I didn't know that :( Seems like it is not safe to persist `tsBuildInfoFile`s 

Very good catch

Comment (sirreal) [reply packages/calypso-build/webpack/sass.js:34]: An alternative I've been meaning to propose is to let TypeScript determine where to put the tsBuildInfoFiles by default. I believe it will put them next to the build output, which makes it easy to cache them all-or-nothing.

We do a lot of manual things that `tsc` has commands for. We don't need to manually clean TypeScript output, for example.

I haven't had an opportunity to apply these improvements to Calypso, but a couple of projects that use this kind of setup:
- Gutenberg: [build](https://github.com/WordPress/gutenberg/blob/a0f36df8a596dc9043cf407a0fd0fdb84490f9aa/package.json#L211) and [clean](https://github.com/WordPress/gutenberg/blob/a0f36df8a596dc9043cf407a0fd0fdb84490f9aa/package.json#L208).
- Gutenberg is quite complex, but I think it's a good model for what the Calypso tsc setup should be like.. Tannin is a smaller project that's easier to grok: https://github.com/aduth/tannin/pull/4

Comment (scinos) [reply packages/calypso-build/webpack/sass.js:34]: Raised https://github.com/Automattic/wp-calypso/pull/45178 for removing custom path for tsBuildInfo files.

**Comment by scinos** [comment]
Blocked by #45178 (or another solution to caching tsBuildInfo files)

File: .circleci/config.yml
Comment (nsakaimbo) [inline .circleci/config.yml:223]: Changes look good, but one more quick question:

This cache key is limited to `CIRCLE_JOB` and `arch`, which won't change between CI jobs/branches/SHAs. It doesn't seem that the cache will ever be invalidated, unless I'm missing something?

In contrast, the key in for Calypso is `v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-build-{{ .Branch }}-{{ .Revision }}`, which at the very least has the revision. Should we use a similar setup here as well?

Comment (scinos) [reply .circleci/config.yml:223]: Ah right, forgot that caches are immutable in CircleCI. Added `.Branch` and `.Revision` as part of the key like Calypso.
Good catch!

File: .circleci/config.yml
Comment (sirreal) [inline .circleci/config.yml:226]: Optional, but I'd consider adding the environment `GLOBAL_CACHE_PREFIX` here as well. For context, some things in the cache seem to have unbounded growth. The environment variable can be managed by CircleCI UI so makes it trivial to flush caches without needing to commit anything.

It makes sense to flush caches periodically because at a certain point it's worth it to flush them once to reduce the size of what we store and retrieve frequently throughout the process.

```suggestion
      - v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-desktop-calypso-build-{{ .Environment.CIRCLE_JOB }}-{{ arch }}-{{ .Branch }}-{{ .Revision }}
      - v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-desktop-calypso-build-{{ .Environment.CIRCLE_JOB }}-{{ arch }}-{{ .Branch }}
      - v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-desktop-calypso-build-{{ .Environment.CIRCLE_JOB }}-{{ arch }}-trunk
      - v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-desktop-calypso-build-{{ .Environment.CIRCLE_JOB }}-{{ arch }}
  desktop-save-calypso-build-cache: &desktop-save-calypso-build-cache
    name: Save Calypso build cache
    key: v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-desktop-calypso-build-{{ .Environment.CIRCLE_JOB }}-{{ arch }}-{{ .Branch }}-{{ .Revision }}
```

Even better would be to automate that to clear caches every month with something like:

```sh
# Run as part of prepare step
date +%Y-%m > ci-cache-prefix.txt
```

Use it as a prefix:
```yml
      - {{ checksum "ci-cache-prefix.txt" }}-desktop-packages-{{ .Environment.CIRCLE_JOB }}-{{ arch }}-{{ checksum "packages-hash" }}
```

File: client/webpack.config.js
Comment (sirreal) [inline client/webpack.config.js:185]: Did you intentionally change the docker exception here? We didn't put the cache into Docker because subsequent Docker images will not benefit from them.

**Review by sirreal — APPROVED** [review]
I left a couple notes, but this looks good to me in general. Thanks for the simplification here 👏

File: client/webpack.config.js
Comment (scinos) [reply client/webpack.config.js:185]: Yes I did, to treat terser cache like all other caches (eg: babel). These are deleted from the Docker image anyway, so it shouldn't affect the size:

```
RUN        WORKERS=$workers CALYPSO_ENV=production BUILD_TRANSLATION_CHUNKS=true yarn run build && rm -fr .cache
```

Nevertheless, I'll keep an eye on Docker image size to make sure it doesn't increase because this change

File: .circleci/config.yml
Comment (nsakaimbo) [reply .circleci/config.yml]: @sirreal Your suggestion makes sense - thanks for pointing this out. However I'm a little confused by the suggested change wrt to the sequence of restoration keys:

```
      - v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-desktop-calypso-build-{{ .Environment.CIRCLE_JOB }}-{{ arch }}-{{ .Branch }}
      - v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-desktop-calypso-build-{{ .Environment.CIRCLE_JOB }}-{{ arch }}-trunk
```

`trunk` is used here while not being part of the longer prefix used prior (either for saving or restoring). My understanding is this will effectively be a no-op as it eventually matches against `- v{{ .Environment.GLOBAL_CACHE_PREFIX }}-v0-desktop-calypso-build-{{ .Environment.CIRCLE_JOB }}-{{ arch }}`. Is the usage of `trunk` here intentional?

Comment (nsakaimbo) [reply .circleci/config.yml]: Edit: I suppose this might be a question for @scinos, as I see the usage of trunk was introduced prior to the suggested change.^

Comment (sirreal) [reply .circleci/config.yml]: It is intentional. `trunk` is replacing `master`. It is (or will be) the default branch and a good assumption is that it's the next best thing if we can't match the branch we're on.

> My understanding is this will effectively be a no-op as it eventually matches against

I don't completely understand your question. These caches find the first matching candidate and use that. So the `…-trunk` candidate will match a recent cache like `…-trunk-abc123`. It's likely that subsequent candidates will not be reached.

**Review by nsakaimbo — APPROVED** [review]
Changes LGTM! Incorporating @sirreal's suggested change to use `GLOBAL_CACHE_PREFIX` sounds like a good idea prior to landing this. Thanks for the awesome work here, @scinos!

**Pull request merged**
